tests/online/cert.c

```diff
@@ -0,0 +1,257 @@
+		git_repository_free(g_repo);
+		g_repo = NULL;
+	}
+	cl_fixture_cleanup("./fake");
+}
+
+static int certificate_check_cb(git_cert *cert, int valid, const char *host, void *payload)
+{
+	struct cert_options *options = payload;
+
+	GIT_UNUSED(cert);
+
+	if (options) {
+		options->had_callback = 1;
+		cl_assert_equal_i(valid, options->expect_valid);
+		if (options->expect_host) {
+			cl_assert_equal_s(host, options->expect_host);
+		}
+	}
+
+	return (options ? options->return_code : GIT_ERROR);
+}
+
+void test_online_cert__certificate_invalid(void)
+{
+	if (!g_has_ssl)
+		cl_skip();
+
+	g_options.fetch_opts.callbacks.certificate_check = certificate_check_cb;
+
+	g_cert_options.expect_valid = 1;
+	g_cert_options.expect_host = "github.com";
+	g_cert_options.return_code = GIT_ECERTIFICATE;
+
+	cl_git_fail_with(GIT_ECERTIFICATE,
+					 git_clone(&g_repo, "https://github.com/libgit2/TestGitRepository", "./fake", &g_options));
+}
+
+void test_online_cert__certificate_invalid_ssh(void)
+{
+	if (!g_has_ssh)
+		cl_skip();
+
+	g_options.fetch_opts.callbacks.certificate_check = certificate_check_cb;
+
+	g_cert_options.expect_valid = 0;
+	g_cert_options.expect_host = "github.com";
+	g_cert_options.return_code = GIT_ECERTIFICATE;
+
+	cl_git_fail_with(GIT_ECERTIFICATE,
+					 git_clone(&g_repo, "ssh://github.com/libgit2/TestGitRepository", "./fake", &g_options));
+
+	cl_assert_equal_i(g_cert_options.had_callback, 1);
+}
+
+void test_online_cert__certificate_valid(void)
+{
+	if (!g_has_ssl)
+		cl_skip();
+
+	g_options.fetch_opts.callbacks.certificate_check = certificate_check_cb;
+
+	g_cert_options.expect_valid = 1;
+	g_cert_options.expect_host = "github.com";
+
+	cl_git_pass(git_clone(&g_repo, "https://github.com/libgit2/TestGitRepository", "./fake", &g_options));
+
+	cl_assert_equal_i(g_cert_options.had_callback, 1);
+}
+
+void test_online_cert__start_with_http(void)
+{
+	if (!g_has_ssl)
+		cl_skip();
+
+	g_options.fetch_opts.callbacks.certificate_check = certificate_check_cb;
+
+	g_cert_options.expect_valid = 1;
+	g_cert_options.expect_host = "github.com";
+
+	cl_git_pass(git_clone(&g_repo, "http://github.com/libgit2/TestGitRepository", "./fake", &g_options));
+
+	cl_assert_equal_i(g_cert_options.had_callback, 1);
+}
+
+void test_online_cert__passthrough_https(void)
+{
+	if (!g_has_ssl)
+		cl_skip();
+
+	g_options.fetch_opts.callbacks.certificate_check = certificate_check_cb;
+
+	g_cert_options.expect_valid = 1;
+	g_cert_options.return_code = GIT_PASSTHROUGH;
+
+	cl_git_pass(git_clone(&g_repo, "https://github.com/libgit2/TestGitRepository", "./fake", &g_options));
+}
+
+void test_online_cert__passthrough_https_bad_certificate(void)
+{
+	if (!g_has_ssl)
+		cl_skip();
+
+	g_options.fetch_opts.callbacks.certificate_check = certificate_check_cb;
+
+	g_cert_options.expect_valid = 0;
+	g_cert_options.return_code = GIT_PASSTHROUGH;
+
+	cl_git_fail_with(GIT_ERROR,
+					 git_clone(&g_repo, "https://wrong.host.badssl.com/fake.git", "./fake", &g_options));
```

Comment:
This should be returning `GIT_ECERTIFICATE` as the passthrough means we want to use the result that the library was going to use if we didn't have the callback.

I believe this `GIT_ERROR`/`-1` actually comes from there not actually being a git repository at the given location.

Reply:
I'll have to recheck what's actually going on w.r.t to the first part. The second part is correct, we're actually succeeding the cert check, but the clone machinery obviously fails down the line, returning `GIT_ERROR`.